src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,325 @@
+        B=None,
+        C=None,
+        backend="cpu",
+    ):
+        if backend == "gpu" and not CUPY_AVAILABLE:
+            raise RuntimeError("Cupy is required for gpu computations.")
+        if backend == "torch":
+            self.xp = torch
+        if backend == "gpu":
+            self.xp = cp
+        elif backend == "cpu":
+            self.xp = np
+        else:
+            raise ValueError("Unsupported backend.")
+        self._fourier_op = fourier_op
+
+        # if not fourier_op.uses_sense:
+        # raise ValueError("please use smaps.")
+
+        # self.n_samples = fourier_op.n_samples
+        self.n_coils = fourier_op.n_coils
+        self.shape = fourier_op.shape
+        self.smaps = fourier_op.smaps
+        self.autograd_available = fourier_op.autograd_available
+
+        if B is not None and C is not None:
+            self.B = self.xp.asarray(B)
+            self.C = self.xp.asarray(C)
+        else:
+            fieldmap = self.xp.asarray(fieldmap)
+            self.B, self.C = get_interpolators_from_fieldmap(
+                fieldmap, t, n_time_segments, n_bins, mask
+            )
+        if self.B is None or self.C is None:
+            raise ValueError("Please either provide fieldmap and t or B and C")
+        self.n_interpolators = len(self.C)
+
+    def op(self, data, *args):
+        """Compute Forward Operation with off-resonance effect.
+
+        Parameters
+        ----------
+        x: numpy.ndarray or cupy.ndarray
+            N-D input image
+
+        Returns
+        -------
+        numpy.ndarray or cupy.ndarray
+            masked distorded N-D k-space
+        """
+        y = 0.0
+        data_d = self.xp.asarray(data)
+        for idx in range(self.n_interpolators):
+            y += self.B[idx] * self._fourier_op.op(self.C[idx] * data_d, *args)
+
+        return y
+
+    def adj_op(self, coeffs, *args):
+        """
+        Compute Adjoint Operation with off-resonance effect.
+
+        Parameters
+        ----------
+        x: numpy.ndarray or cupy.ndarray
+            masked distorded N-D k-space
+
+        Returns
+        -------
+            inverse Fourier transform of the distorded input k-space.
+        """
+        y = 0.0
+        coeffs_d = self.xp.array(coeffs)
+        for idx in range(self.n_interpolators):
+            y += self.xp.conj(self.C[idx]) * self._fourier_op.adj_op(
+                self.xp.conj(self.B[idx]) * coeffs_d, *args
+            )
+
+        return y
+
+    def get_grad(self, image_data, obs_data):
+        """Compute the data consistency error.
+
+        Parameters
+        ----------
+        image_data: numpy.ndarray or cupy.ndarray
+            N-D input image
+        obs_data: numpy.ndarray or cupy.ndarray
+            N-D observed k-space
+
+        Returns
+        -------
+        numpy.ndarray or cupy.ndarray
+            data consistency error in image space.
+        """
+        return self.adj_op(self.op(image_data) - obs_data)
```

Comment:
We call this `data_consistency` and we already have it defined, this is not needed.
```suggestion
```

Reply:
I think this was already in the code (perhaps some initial version before data_consistency was defined?). I agree this is redundant now!